### DISCUSSION TITLE
How to use multi-root directed graphs

### DISCUSSION THREAD

**Discussion by hombit** [egraphs-good/egg#212] [Q&A] [answered]
I have a problem to optimize a total weight of a positive-weight directed graph. The solution must keep all roots and leaves nodes, while maintaining them connected. Could I solve such a problem with the `egg`?

A simplest example of what I want is to replace this
```mermaid
graph TD;
    A -- 1 --> B;
    A -- 2 --> C;
    B -- 1 --> D;
    C -- 1 --> D;
```
with
```mermaid
graph TD;
    A -- 1 --> B;
    B -- 1 --> D;
```

Could you please help me to figure out how to build such graphs with `egg` and what kind of rules I need?

**Comment by yihozhang** [comment, marked as answer]
This sounds like a hard problem. It seems to me that the [Steiner tree problem](https://en.wikipedia.org/wiki/Steiner_tree_problem), which is NP-hard, can be viewed as an instance of this problem where the roots and the leaves contain the same set of vertices and the graph is undirected.

**Reply by hombit** [reply]
Thank you! Graph theory is new to me, and I believe you've given a right direction to me

**Comment by omidmohammdi** [comment]
I think you can solve this problem with the Steiner tree when your choice is based on the candidate node or in other words selective about the middle nodes.
